src/serialization/json/JsonSerializationWriter.cs

```diff
@@ -360,6 +361,33 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
             }
         }
         /// <summary>
+        /// Writes the specified dictionary to the stream with an optional given key.
+        /// </summary>
+        /// <param name="key">The key to be used for the written value. May be null.</param>
+        /// <param name="values">The dictionary of values to be written.</param>
+#if NET5_0_OR_GREATER
+        public void WriteDictionaryValue<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] T>(string? key, T values) where T : IDictionary
+#else
+        public void WriteDictionaryValue<T>(string? key, T values) where T : IDictionary
+#endif
+        {
+            if(values != null)
+            {
+                if(!string.IsNullOrEmpty(key))
+                    writer.WritePropertyName(key!);
+                var type = values.GetType();
+                if(!type.IsGenericType || type.GetGenericArguments()[0] != typeof(string))
+                    throw new InvalidOperationException($"error serialization additional data value with key {key}, unsupported type {type}");
+
+                writer.WriteStartObject();
+                foreach(DictionaryEntry entry in values)
+                {
+                    WriteAnyValue((string)entry.Key, entry.Value);
+                }
```

Comment:
```suggestion
                writer.WriteStartObject();
                foreach(DictionaryEntry entry in values)
                {
                    if (entry.Key is not string keyStr)
                        throw new InvalidOperationException($"error serialization additional data value with key {key}, unsupported type {type}");
                    WriteAnyValue(keyStr, entry.Value);
                }
```

While functional, advanced type reflection has a high runtime cost and is going to trip AOT/trimming

Reply:
Happy to iterate on this.
Is there something specific you would rather I avoid?

Reply:
yes, we should avoid using reflection as much as possible here since it throws AOT/trimming off and is costly at runtime.
I believe the suggestion I've made is functionally equivalent to the initial proposal here.

Reply:
Ah sorry I missed the suggestion, yeah that works for me.

Reply:
We still need to use reflection for the error message but this is inline with the WriteAnyValue so I assume that's fine.

Reply:
I will look at adding a test for the exception too.